### DISCUSSION TITLE
How to resolve the issue of an entity not being fully displayed after adding it and rotating the view?

### DISCUSSION THREAD

**Discussion by MYDJH** [iTwin/itwinjs-core#6850] [Q&A] [answered]
Hello, I am aiming to achieve a flood simulation effect using iTwin.js, similar to the ocean effect in three.js: https://threejs.org/examples/?q=water#webgl_shaders_ocean.

My initial approach was to add a blue rectangular box to the model and control its height to simulate rising and falling water levels. However, after adding the blue box to my model, I noticed that when I rotate the model, the box is not fully visible. I want it to be visible from any angle, just like my model. Is this due to the camera settings?

Here is the reference code I used to add the box: https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%203d
Here is a screenshot of the issue:
![%H_O8PI3{A8Z@R72MZ88SU3](https://github.com/iTwin/itwinjs-core/assets/101549988/ae1fa7cf-1bd1-4f5f-a25a-14cba526880c)
![$)%D3~CZJ)1{A8)W(@S_5XU](https://github.com/iTwin/itwinjs-core/assets/101549988/8192c598-08b4-4bd9-b354-d49a1ce0d049)
![{4%WO%R$ESPL$AF0 G)1 CC](https://github.com/iTwin/itwinjs-core/assets/101549988/3d4aaa19-f1e2-4fd6-b821-348508b8a373)

Do you have any alternative approaches or suggestions? I appreciate your help and advice.

**Comment by pmconne** [comment]
The view's near and far planes are computed from the bounding boxes of the geometry in the view, but this doesn't currently include decorators. So your decoration geometry can get clipped when it extends outside of the iModel's bounding box.

If you were drawing your geometry as a tile tree instead of a decorator, its bounding box would be included in the computation of the view frustum. And IMO the kind of geometry you are trying to draw (something that's supposed to be part of the scene, instead of just "decorating" it) should be drawn as a tile tree.

You can very easily create a tile tree from a graphic and add it to the scene using [TileTreeReference.createFromRenderGraphic](https://www.itwinjs.org/reference/core-frontend/tiles/tiletreereference/createfromrendergraphicstatic/).

**Reply by MYDJH** [reply]
Thank you very much for your suggestion.
 Unfortunately, I encountered another issue when trying this method.
 I found that using const treeRef = TileTreeReference.createFromRenderGraphic requires a higher version of the @itwin/core-frontend package. My project currently uses "@itwin/core-frontend": "^4.3.3". When I try to install a newer version, compatibility issues arise with other packages like @itwin/appui-abstract, preventing me from proceeding. I tried manually updating the conflicting files, such as running npm install @itwin/core-frontend@4.7.1, but I found that reinstalling one package caused conflicts with others.
I want to update the dependencies in my project. How can I determine a stable and non-conflicting set of version dependencies? Do you have any good suggestions?
Here is the version dependency file for my project：
[package.json](https://github.com/user-attachments/files/15885590/package.json)

**Reply by MYDJH** [reply]
This is the actual version of the project when I run npm list.If you have any good suggestions, I would greatly appreciate it.
[list.txt](https://github.com/user-attachments/files/15885777/list.txt)

**Reply by pmconne** [reply]
It will help if you provide details about how you went about attempting to update and what specific errors you encountered.

@GerardasB @aruniverse.

**Reply by MYDJH** [reply]
Hello ,Here is a detailed description of the problem.
When I execute this code, according to the TypeScript hint, I find that it lacks the createFromRenderGraphic method.
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/6afd742d-ed63-4bc5-88e7-9f20bf339f2d)

I found that my TileTreeReference class does not have the createFromRenderGraphic method, but this method exists in the latest version of @itwin/core-frontend.
I tried using npm i @itwin/core-frontend@4.7.1.
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/d0faf463-5832-4fa1-9a99-7523ce0cde98)

I also tried npm i @itwin/core-frontend@4.5.0, but it also reported conflicting files.
I tried removing it with npm uninstall @itwin/core-frontend and then reinstalling, but issues still occur.
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/55c734c2-ddad-45ed-b440-60dfa0ca45e1)

I also tried manually resolving conflicts as per the peer suggestions mentioned above. When I removed conflicting files, upgraded, and installed the latest dependencies, I encountered conflicts with other packages again. I want to test to find a version that allows me to use the createFromRenderGraphic method.

I want to upgrade all my dependencies comprehensively, but I haven't found a suitable version. The only thing I can think of is testing each version one by one, but that feels like a lot of work. I'm wondering if there's a better suggestion.

**Reply by divyashah0510** [reply]
@MYDJH the issue might be with the version specification if you try with npm i @itwin/core-frontend the issue would be solved

**Reply by aruniverse** [reply]
iTwinjs-Core packages are all in lockstep, so you can't just install core-frontend@4.7.1 and leave the rest on 4.3.3, all the core packages would need to be updated to 4.7.1. That should eliminate the conflicting peer deps for appui-abstract.

**Reply by MYDJH** [reply]
Thank you very much, the issue has indeed been resolved.

**Comment by MYDJH** [comment]
@pmconne 
Hello, I'm sorry to trouble you again. I've tried using TileTreeReference.createFromRenderGraphic to add a rectangle to the view, but I found that when I rotate the view, some parts are not displayed fully. It only displays normally when its size is consistent with the size of my model. 
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/4bf7e063-9a0b-4653-b5f7-2266e60bb3d0)

Whenever I zoom in, it causes the display to be incomplete. I suspect this might be related to the frustum, so I made some modifications, but I failed.
This is my code：
```
  function AddBoxMethod() {
    const viewport = IModelApp.viewManager.selectedView;
    if (viewport) {
      addTiledGraphics(viewport);
      setCustomView(viewport);
    }
    function addTiledGraphics(viewport: Viewport): void {
      const builder = IModelApp.renderSystem.createGraphic({
        type: GraphicType.Scene,
        computeChordTolerance: () => 0.01,
      });
      const rangeCompare = viewport.iModel.projectExtents;
      // increase rangeCompare by two times
      let copyRangeBox = rangeCompare.clone();
      const scale = 2;
      const center = copyRangeBox.center;
      const delta = center.vectorTo(copyRangeBox.high).scale(scale - 1);
      copyRangeBox.low.subtractInPlace(delta);
      copyRangeBox.high.addInPlace(delta);

      builder.addSolidPrimitive(Box.createRange(copyRangeBox, true)!);
      const graphic = builder.finish();
      const treeRef = TileTreeReference.createFromRenderGraphic({
        graphic,
        modelId: viewport.iModel.transientIds.getNext(),
        iModel: viewport.iModel,
      });
      viewport.addTiledGraphicsProvider({
        forEachTileTreeRef: (_vp, func) => func(treeRef),
      });
    }
    function setCustomView(viewport: Viewport): void {
      const range2 = viewport.iModel.projectExtents;
      let copyRange = range2.clone();
      const center = copyRange.center;
      const scale = 3;
      const delta = center.vectorTo(copyRange.high).scale(scale - 1);
      copyRange.low.subtractInPlace(delta);
      copyRange.high.addInPlace(delta);
      const frustum = viewport.getFrustum();
      frustum.setFrom(Frustum.fromRange(copyRange!));
      viewport.setupViewFromFrustum(frustum);
      viewport.changeView(viewport.view);
      viewport.view.lookAtVolume(copyRange!, viewport.viewRect.aspect);
      viewport.synchWithView({ animateFrustumChange: true });
    }
  }

```
"This is the problematic image. Could you help me analyze what is causing this issue? Thank you."
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/54d506e9-3f2b-4359-8499-74fffa8979a2)
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/62f59fe1-2500-40f8-a3f1-f01563e04c4c)
![image](https://github.com/iTwin/itwinjs-core/assets/101549988/50c6d555-4820-4780-81de-c116afc90322)

**Reply by pmconne** [reply]
Thank you. I see the issue. Filed as #6873. Should be an easy fix.

**Reply by pmconne** [reply, marked as answer]
Fixed by #6876.

**Reply by pmconne** [reply]
Fix will be included in 4.7.2. Thanks for your help.

**Reply by pmconne** [reply]
4.7.2 has been released.

**Reply by MYDJH** [reply]
Yes, the issue is resolved now. Thank you very much for your help!
